### PR TITLE
fix CUDA6.5 int(bool) bug in particle kernel

### PR DESCRIPTION
**Pull request by PrometheusPi**
This pull request solves the CUDA6.5 `int(bool)` bug (#570) in `Particles.kernel`.

In order to work with the bool arrays of size `simDim`, `for` loops were used. These also avoid using *ugly* pre-compiler `#if`.

**TO DOs:**
 - [x] verify that this change does not slow done PIConGPU
 - [x] varify that particles are still transported correctly

### CODE REVIEW

File: src/picongpu/include/particles/Particles.kernel [hunk 1]
```diff
@@ -517,13 +518,13 @@ struct PushParticlePerFrame
...
-            + (dir.z() + int(dir.z() < 0)*3) * BACK
-#endif
-            ;
+        int direction = 1;
+        uint32_t exchangeType = 1; // see libPMacc/include/types.h for RIGHT, BOTTOM and BACK
+        for (uint32_t i = 0; i < simDim; ++i)
+        {
+            direction += (dir[i] + (dir[i] < 0 ? 3 : 0)) * exchangeType; 
+            exchangeType *= 3; // =3^i
+        }
```
Comment (PrometheusPi) [inline src/picongpu/include/particles/Particles.kernel]: @ComputationalRadiationPhysics/picongpu-maintainers 

Do you think this is cleaner/easer to read than:
```c++
const int direction = 1 +
    (dir.x() + int(dir.x() < 0)*3) * RIGHT 
    + (dir.y() + int(dir.y() < 0)*3) * BOTTOM
#if (SIMDIM==DIM3)
    + (dir.z() + int(dir.z() < 0)*3) * BACK
#endif
    ;
```

File: src/picongpu/include/particles/Particles.kernel [hunk 2]
```diff
@@ -489,7 +489,8 @@ struct PushParticlePerFrame
          * if particle is inside of the supercell the **unsigned** representation
```
Comment (ax3l) [inline src/picongpu/include/particles/Particles.kernel]: above: outdated comments "ATTENTION ATTENTION we cast to unsigned ..." ?

Comment (ax3l) [inline src/picongpu/include/particles/Particles.kernel]: ah no, still up to date

File: src/picongpu/include/particles/Particles.kernel [hunk 3]
```diff
@@ -489,7 +489,8 @@ struct PushParticlePerFrame
...
          * of dir is always >= size of the supercell
          */
-        dir = precisionCast<int>(precisionCast<uint32_t >(localCell) >= precisionCast<uint32_t >(TVec::toRT())) * dir;
```
Comment (ax3l) [inline src/picongpu/include/particles/Particles.kernel]: that is *not* the bug signature you are looking for.

Comment (ax3l) [inline src/picongpu/include/particles/Particles.kernel]: too many parenthesis, you are right

**Comment by PrometheusPi** [comment]
There is **no performance reduction** due to this fix:

```bash
$ for i in `seq 10 10 100`; do grep "$i %" 00[8,9]*/simOutput/output; echo "-------------" ; done
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 10 % =      200 | time elapsed:       1min 11sec 704msec | avg time per step: 336msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 10 % =      200 | time elapsed:       1min 11sec 604msec | avg time per step: 336msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 20 % =      400 | time elapsed:       2min 26sec 471msec | avg time per step: 343msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 20 % =      400 | time elapsed:       2min 26sec 634msec | avg time per step: 343msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 30 % =      600 | time elapsed:       3min 42sec 121msec | avg time per step: 346msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 30 % =      600 | time elapsed:       3min 42sec 406msec | avg time per step: 347msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 40 % =      800 | time elapsed:       4min 58sec 715msec | avg time per step: 349msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 40 % =      800 | time elapsed:       4min 59sec  40msec | avg time per step: 350msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 50 % =     1000 | time elapsed:       6min 16sec 144msec | avg time per step: 356msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 50 % =     1000 | time elapsed:       6min 16sec 587msec | avg time per step: 357msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 60 % =     1200 | time elapsed:       7min 35sec 814msec | avg time per step: 369msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 60 % =     1200 | time elapsed:       7min 36sec 377msec | avg time per step: 370msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 70 % =     1400 | time elapsed:       9min  2sec  30msec | avg time per step: 410msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 70 % =     1400 | time elapsed:       9min  2sec 800msec | avg time per step: 411msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 80 % =     1600 | time elapsed:      10min 36sec 589msec | avg time per step: 444msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 80 % =     1600 | time elapsed:      10min 37sec 560msec | avg time per step: 445msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>: 90 % =     1800 | time elapsed:      12min 12sec 946msec | avg time per step: 448msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>: 90 % =     1800 | time elapsed:      12min 14sec  64msec | avg time per step: 448msec
-------------
008_fix_particleKernel/simOutput/output:[1,0]<stdout>:100 % =     2000 | time elapsed:      13min 50sec 137msec | avg time per step: 450msec
009_dev_particleKernel/simOutput/output:[1,0]<stdout>:100 % =     2000 | time elapsed:      13min 51sec 465msec | avg time per step: 451msec
-------------
```

**Comment by PrometheusPi** [comment]
There is also **no numerical heating**:
particlekernel_numheating

File: src/picongpu/include/particles/Particles.kernel (hunk 1 above)
Comment (ax3l) [inline src/picongpu/include/particles/Particles.kernel]: :)
i_see_magic_numbers

Comment (PrometheusPi) [inline src/picongpu/include/particles/Particles.kernel]: Me too. Any idea how to use the *magic numbers* from [`libPMacc/include/types.h`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f0051b7f1c33a716fcee8cf91657f78b96e0bffd/src/libPMacc/include/types.h#L90-L93) in an iterate-able way?

Comment (psychocoderHPC) [inline src/picongpu/include/particles/Particles.kernel]: we can the comparison  to the same like the comment
`direction += (dir[i] + (dir[i] < 0 ? 3 : 0)) * exchangeType; `
to
`direction += (dir[i] == -1 ? 2 : dir[i]) * exchangeType; `

**Comment by PrometheusPi** [comment]
**I still need to check if there is no numerical heating and no slow down due to the last change.**

@ax3l and @psychocoderHPC  Should I combune both commits to one?

**Comment by PrometheusPi** [comment]
Heating stayed the same and there is no slow down.

**Pull request merged**